<h1 style='font-size:40px'> Onwards to Kubernetes!</h1>

<h2 style='font-size:30px'> 
    The Why's and What's of Kubernetes</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            O Kubernetes é uma tecnologia de escalabilidade de containers. Quando uma determinada instância é sobrecarregada, o software pode criar uma nova.
        </li>
        <li>
            O software cria máquinas virtuais (ou nós), que hospedarão múltiplos containers. As funcionalidades de cada nó são gerenciadas pela entidade Master.
        </li>
        <li>
            A agregação de nós com um Master é conhecida como Cluster.
            <center style='font-size:20px;margin-top:20px'> 
                 <img src='img/12_kubernetes_diagram.png'>
            </center>
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
    Kubernetes in Development and Production</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Lidar com clusters Kubernetes no ambiente de dev ou prod requer soluções diferentes. Em nossa máquina local, usamos o minikube para gerenciar os nós; já em ambientes de nuvem, costumamos usar a solução do provedor (EKS).
        </li>
        <li>
            Devemos ainda contar com outro software (o kubectl) para o gerenciamento dos containers dos nós.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
    Mapping Existing Knowledge</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Esta seção será dedicada ao deployment da aplicação feita na parte 10 do curso com Kubernetes.
        </li>
        <li>
            Vale mencionar que o Kubernetes possui algumas particularidades: todas as imagens deverão já estar construídas e a rede de containers/clusters terá que ser manualmente criada por nós.
        </li>
        <li>
            Com relação às imagens, elas deverão estar publicadas no Docker Hub. O Kubernetes não reconhece imagens na máquina local.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
    Object Types and API Versions</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Todos os arquivos de configuração do Kubernetes parametrizam um objeto dentro de nosso ambiente. O tipo de objeto a ser definido é especificado no argumento `kind`.
        </li>
        <li>
            O tipo de objeto ao qual vamos ter acesso dependerá da apiVersion que especificarmos em nosso .yaml.
        </li>
        <li>
            Objetos Pod são responsáveis por rodar containers; já os de Service, por estabelecer uma rede.
        </li>
    </ul>
 </div>

<figure>
        <center style='margin-top:20px'>
            <img src='img/12_object_kind.png'>
                <figcaption>  Seguem alguns objetos admitidos pelo Kubernetes (apiVersion: v1).</figcaption>
        </center>
    </figure>

<h2 style='font-size:30px'> 
    Running Containers in Pods</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Aula em que explicamos de maneira mais detalhada o papel de um Pod.
        </li>
        <li>
            Pods são objetos em que deployamos containers com alta interdependência. Isso significa que o desligamento de um deles inutilizaria os demais.
        </li>
    </ul>
 </div>

In [5]:
! git add .
! git commit -m 'Aula 185'
! git push

[master fea67ce] Aula 184
 2 files changed, 44 insertions(+), 26 deletions(-)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 24 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 842 bytes | 842.00 KiB/s, done.
Total 4 (delta 3), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (3/3), completed with 3 local objects.
To https://github.com/felipesveiga/docker.git
   09d743c..fea67ce  master -> master


<p style='color:red'> Vi Aula 184; Aula 185</p>

<p style='color:red'> Para criar um cluster, utilize o comando `minikube start --driver=docker`</p>